src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java

```diff
@@ -352,6 +351,8 @@ private void notifyEnd(DetailAST rootAST, AstState astState) {
 
         for (AbstractCheck check : checks) {
             check.finishTree(rootAST);
+            addMessages(check.getMessages());
+            check.clearMessages();
```

Comment:
@romani If any check has exception in `finishTree` or `visitToken`, messages will never be cleared and appear in the next file run. We saw something like this in sevntu. https://github.com/sevntu-checkstyle/sevntu.checkstyle/issues/589 .
IMO, we should clear messages right before `beginTree`.
We are already doing this in `AbstractFileSetCheck` in this PR.